2단계 - 문자열 덧셈 계산기

hyeok7777 opened the pull request.

hyeok7777 committed:
1단계 - 학습 테스트 실습

hyeok7777 committed:
2단계 - 문자열 덧셈 계산기

hyeok7777 committed:
refactor : comment 반영

hyeok7777 committed:
Merge branch 'hyeok7777' into hyeok7777-step2

hyeok7777 commented:
늦은 시간이라 주무실 줄 알고 브랜치를 지우지 않고 새로 브랜치를 따서 작업하다가 커밋이 좀 꼬였습니다. 🙁 다음 단계부터는 제대로 진행하겠습니다. 
추가적으로 3단계 미션부터는 커밋 단위 분리해서 진행할 수 있도록 하겠습니다.

wooobo commented on src/main/java/MakeArraySum.java:
```diff
@@ -0,0 +1,10 @@
+
+    public int getSum(String[] numbers) {
+        int sum = 0;
+        for(int i = 0 ; i < numbers.length ; i++) {
```
`포맷팅` 이 적용안되있어요~
`for (int i = 0; i < numbers.length; i++) {`

`IDE`에 `Google Java Convention ` 파일을 적용하셔서 
`IDE` 별로 자동 포맷팅 기능이 있으니 찾아보시고 저장시 마다 자동 포매팅 설정하는것이 좋습니다!

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,53 @@
+
+public class NumberValidator {
+    private static final String CUSTOM_REGEX = "//(.)\n(.*)";
+    private static MakeArraySum makeArraySum = new MakeArraySum();
```
`MakeArraySum` 클래스에서 `클래스 변수`가 존재하지 않습니다.
그렇다면, 
`private static MakeArraySum makeArraySum = new MakeArraySum();`
보다는 `static` 메소드를 사용해 보는건  어떨까요?

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,53 @@
+    private String[] checkRegex() {
+        Matcher m = getMatcher();
+        if (m.find()) {
+            String customDelimiter = m.group(1);
```
특정 숫자를 직접 사용하는것보다는 상수로 선언해보는 고민을 해보시면 좋을것같아요!

[매직넘버 치환](https://hoonmaro.tistory.com/44)

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private Matcher getMatcher() {
+        Matcher m = Pattern.compile(CUSTOM_REGEX).matcher(text);
```
`Pattern` 인스턴스 생성비용이 비싼편입니다.

현재 미션에서는 `Pattern`을 과도하게 사용하지는 않지만,
많은 호출이 될 경우, 매번 `Pattern` 인스턴스를 생성하기 보다는 재사용성을 높이기 위해 캐싱전략을 해보시면 어떨까요? [참고](https://icarus8050.tistory.com/64)



----

m 변수에대한 사용하는 부분이 없다면
변수선언 없이 바로 return하는 편이 더 좋지 않을까요? :thinking:

불필요한 변수선언은 시간이 지날수록 코드의 복잡도를 증가실수 있다고 생각해요
추가적으로 모든 변수명에 의미있는 이름을 지어주는것도 중요합니다!

wooobo commented on README.md:
```diff
@@ -25,3 +25,19 @@
+## 2단계
+### 문자열 덧셈 계산기를 통한 TDD/리팩토링 실습
+
+#### 기능 요구사항
```
요구사항 정리해주셨는데요 :+1: 

단순한 작업내용을 작성하는것으로 끝나는것이 아닌

작업시작 전 `todo list`를 작성하여 완료된 목록을 체크 하나하나 나간다면 작업 진행사항을 명확하게 알수 있을거라 생각되어요~
그리고 작업을 진행하면서 추가적으로 진행해야할 작업사항이 보이면 기억으로만 남겨 놓는것이 아닌
기록으로 미리 정리 해놓는다면 혹시 모를 누락사항을 방지할수도 있지않을까 생각합니다 :pray:

예시)
```
- [X] null 을 입력 받을 경우 0을 리턴한다.
- [ ] 숫자하나를 입력 할 경우 숫자 합은 입력 숫자 그대로 이다.

## 리팩토링 작업

- [x] 제공된 테스트 코드에 맞춰서 기능 리팩토링
```

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.regex.Pattern;
+
+public class NumberValidator {
+    private static final String CUSTOM_REGEX = "//(.)\n(.*)";
```
상수선언 :+1: 

다른 `하드코딩`된 코드도 찾아보시고 상수화하시면 좋을것 같아요 :pray:

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,53 @@
+
+    private void checkNegative(String s) {
+        if (s.contains("-")) {
+            throw new RuntimeException();
```
`Exception`을 발생했을때 적절한 예외메시지도 포함해서 `Exception`을 처리한다면 좋지않을까요?
예를 들면... `0이상의 숫자만 입력할수있습니다.` 이런식으로요~

wooobo reviewed (changes requested):
안녕하세요 진혁님~

2단계 미션 잘 작성해주셨습니다 :+1: 

몇가지 소소한 코멘드 남겨두었습니다!
코멘드 참고하셔서 개선해야 될 부분 수정해 보시면 좋을것 같아요! 

리뷰 재요청 부탁드리겠습니다!  :pray:

hyeok7777 committed:
refactor : MakeArraySum.java 포맷팅

hyeok7777 committed:
refactor : todo list로 README.md 변경

hyeok7777 committed:
refactor : NumberValidator.java
1) MakeArraySum static 메소드로 호출
2) 매직넘버 치환
3) Matcher 변수선언 제외 후 return
4) Exception 메세지 수정
5) regex 하드코딩 제거

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -2,8 +2,10 @@
-    private static MakeArraySum makeArraySum = new MakeArraySum();
+    private static final int STATIC_ONE = 1;
+    private static final int STATIC_TWO = 2;
     String text;
```
접근제어자가 빠져있었네요 :sunglasses:

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -19,29 +21,28 @@ public int getNumber() {
         if (m.find()) {
-            String customDelimiter = m.group(1);
-            String[] numbers = m.group(2).split(customDelimiter);
+            String customDelimiter = m.group(STATIC_ONE);
```
상수화 적용 :+1: 

약간 아쉬운점은 매직넘버를 상수화 한다는것은 단순한 숫자로 보여지는것이 아닌 의미있는 객체를 부여해준다라고도 볼수 있을것 같아요.
그런 측면에서  `STATIC_ONE` 보다는 `DELIMITER_INDEX` 이렇게도 해볼수 있을것같아요

의미있는 변수명 선언은 함께하는 개발자 분들이 코드를 이해하는데에도 큰 도움이 됩니다 :pray:

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,54 @@
+        if (text == null || text.equals("")) {
+            return true;
+        }
+        return false;
```
`return text == null || text.equals("")` 이렇게도 가능하겠네요!

개인적으로 `if` 가 많아지면 코드읽는데에 시간이 더 걸린다고 생각해요!

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,54 @@
+
+    private void checkNegative(String s) {
+        if (s.contains("-")) {
+            throw new RuntimeException("0이상의 숫자만 입력할 수 있습니다. (음수 입력 불가)");
```
추가적으로 `RuntimeException` 보다는 의미있는  `Exception` 있을것 같아요 :)

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,54 @@
+            return sum;
+        }
+        String[] numbers = checkRegex();
+        for (int i = 0; i < numbers.length; i++) {
+            checkNegative(numbers[i]);
+        }
+        sum = MakeArraySum.getSum(numbers);
+        return sum;
```
`Arrays.stream` 을 활용해보는 방법을 고민해보아요 :)

wooobo commented on src/main/java/NumberValidator.java:
```diff
@@ -0,0 +1,54 @@
+        this.text = text;
+    }
+
+    public int getNumber() {
```
함수 이름은 `getNumber` 이지만 합산하는 책임까지 부여되어서 적절하지 않는 네이밍인것 같아요
그렇다는건, 책임을 좀 더 분리시켜 보면 어떨까요?
`합산` 로직과 `숫자 검증` 로직의 책임을 분리시키는 부분을 고민해보시면 다음 미션에서 도움 되실꺼에요! 

예를들면,
- `SplitDelimiter` : 문자열을 분리시키는 클래스
- `Number` : 숫자를 검증하는 클래스 (모든 원시값과 문자열을 포장한다)
- `Numbers` : Number의 일급 콜렉션 클래스

wooobo reviewed (approved):
2단계 미션 요청사항 잘 반영해주셨습니다!


추가적은 코멘트 남겨드렸어요~
다음 단계 진행시 함께 고민해보시면 감사드리겠습니다 🙏 

그럼 2단계 미션 `merge` 하겠습니다!

드디어! 본격 로또 미션 시작해보시죠 👍
